# Welcome to lab_🐼s!

Last week, you did amazing work in setting up your computer to do Data Science!  Let's now do Data Science on your computer for the first time!

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant, your Course Aides, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to **communicate about your work**.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to **explain your code**, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- **Every lab** you will need to complete some parts of the lab **with your group**.  There will be both **"group discussions"** and **"individual reflections"** as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [ ]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Kim"
ca1_name = "Jai"
ca2_name = "Mia"


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "Archit"
partner1_netid = "architk2"
partner1_major = "Physics"

partner2_name = "Kasra"
partner2_netid = "kvak12"
partner2_major = "undeclared"

partner3_name = "Brian"
partner3_netid = "bmarti64"
partner3_major = "Civil Engineering "


<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

In lecture, we have already began to explore the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

In this lab, **you** will come up with questions to ask the entire class for the **"Hello Dataset"**!

## Hello Dataset Overview

Soon after this lab, your homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **NOT be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything at most PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*

**What is your binary question?**

*Do you prefer winter or summer?*

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**

*It has two set answers that are different enough to polarize the classes opinions. Winter and summer are complete opposites so we hope this will encourage that split.*

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

*How many classes are you taking?*

**What do you think the most frequent answer will be?**

*We predict the most frequent answer to be 4 classes *

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentences on why you believe that it will be the most frequent answer.**

We said this because the range of credit hours is usually 14-16 which usually corresponds with 4 courses.

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

*What is your favorite holiday?

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

We were split between Christmas and Halloween. I prefer Halloween.

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use one of the first three questions but you want to know the classes' answer?

**What is your final question for the whole class?**

What school year are you?

## Submit Your Questions to Your TA

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/yMUDTbw9FSpEQ6DF7


<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In literally every lab, you will need to import the library just like we did in lecture -- import pandas as `pd` in the following cell:

In [1]:
# Write the code to import pandas as pd:
import pandas as pd

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

Remember, URLs passed to `read_csv` are **string** variables, so remember to use **quotation marks**. 

In [2]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv")

In [3]:
# Print out the contents of `df`:
df

Year    Term YearTerm Subject  Number                          Name  \
0      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
1      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
2      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
3      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
4      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
...     ...     ...      ...     ...     ...                           ...   
11816  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11817  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11818  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11819  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11820  2023  Spring  2023-sp    ZULU     202            Elementary Zulu II   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
11816                    Same as INFO 303. See INFO 303.     3 hours.   
11817                    Same as INFO 303. See INFO 303.     3 hours.   
11818                    Same as INFO 303. See INFO 303.     3 hours.   
11819                    Same as INFO 303. See INFO 303.     3 hours.   
11820  Continuation of ZULU 201 with introduction of ...     5 hours.   

                                            Section Info  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
11816                    Same as INFO 303. See INFO 303.   
11817                    Same as INFO 303. See INFO 303.   
11818                    Same as INFO 303. See INFO 303.   
11819                    Same as INFO 303. See INFO 303.   
11820  Same as AFST 252. Participation in the languag...   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
11816                       Advanced Composition course.  ...              A   
11817                       Advanced Composition course.  ...              A   
11818                       Advanced Composition course.  ...              A   
11819                       Advanced Composition course.  ...              A   
11820                                                NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
0                   Open  Discussion/Recitation       DIS   09:00 AM   
1                   Open  Discussion/Recitation       DIS   10:00 AM   
2                   Open  Discussion/Recitation       DIS   11:00 AM   
3                   Open  Discussion/Recitation       DIS   12:00 PM   
4                   Open  Discussion/Recitation       DIS   01:00 PM   
...                  ...                    ...       ...        ...   
11816  

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [4]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 11821 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [6]:
df_STAT = df[df["Subject"] == "STAT"]
df_STAT

Year    Term YearTerm Subject  Number                           Name  \
11086  2023  Spring  2023-sp    STAT     100                     Statistics   
11087  2023  Spring  2023-sp    STAT     100                     Statistics   
11088  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
11089  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
11090  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
11234  2023  Spring  2023-sp    STAT     590  Individual Study and Research   
11235  2023  Spring  2023-sp    STAT     593                STAT Internship   
11236  2023  Spring  2023-sp    STAT     593                STAT Internship   
11237  2023  Spring  2023-sp    STAT     595       Preparing Future Faculty   
11238  2023  Spring  2023-sp    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11086  First course in probability and statistics at ...        3 hours.   
11087  First course in probability and statistics at ...        3 hours.   
11088  Data Science Discovery is the intersection of ...        4 hours.   
11089  Data Science Discovery is the intersection of ...        4 hours.   
11090  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
11234  Directed reading and research. Approved for le...   0 TO 8 hours.   
11235  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11236  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11237  Prepares Ph.D. students who are interested in ...        2 hours.   
11238  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11086  Credit is not given for both STAT 100 and any ...   
11087  Credit is not given for both STAT 100 and any ...   
11088                         Same as CS 107 and IS 107.   
11089                         Same as CS 107 and IS 107.   
11090                         Same as CS 107 and IS 107.   
...                                                  ...   
11234  Approved for letter and S/U grading. May be re...   
11235  Approved for letter and S/U grading. Prerequis...   
11236  Approved for letter and S/U grading. Prerequis...   
11237                                                NaN   
11238  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11086  Quantitative Reasoning I course.  ...              A   
11087  Quantitative Reasoning I course.  ...              A   
11088  Quantitative Reasoning I course.  ...              A   
11089  Quantitative Reasoning I course.  ...              A   
11090  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
11234                               NaN  ...              A   
11235                               NaN  ...              A   
11236                               NaN  ...              A   
11237                               NaN  ...              A   
11238                               NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11086  Open (Restricted)                Lecture       LEC   10:00 AM   
11087             Closed                 Online       ONL   ARRANGED   
11088      CrossListOpen                Lecture       LEC   12:00 PM   
11089             Closed  Laboratory-Discussion       LBD   03:30 PM   
11090             Closed  Laboratory-Discussion       LBD   03:30 PM   
...                  ...                    ...       ...        ...   
11234             Closed      Independent Study       IND   ARRANGED   
11235  Open (Restricted)             Internship       INT   ARRANGED   
11236  Open (Restri

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [7]:
df_CS = df[df["Subject"] == "CS"]
df_CS

Year    Term YearTerm Subject  Number                              Name  \
3040  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3041  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3042  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3043  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3044  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3373  2023  Spring  2023-sp      CS     598                    Special Topics   
3374  2023  Spring  2023-sp      CS     598                    Special Topics   
3375  2023  Spring  2023-sp      CS     598                    Special Topics   
3376  2023  Spring  2023-sp      CS     598                    Special Topics   
3377  2023  Spring  2023-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3040  Fundamental principles, concepts, and methods ...        3 hours.   
3041  Fundamental principles, concepts, and methods ...        3 hours.   
3042  Fundamental principles, concepts, and methods ...        3 hours.   
3043  Fundamental principles, concepts, and methods ...        3 hours.   
3044  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3373  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3374  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3375  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3376  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3377    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3040  Prerequisite: One of MATH 220 or MATH 221 or M...   
3041  Prerequisite: One of MATH 220 or MATH 221 or M...   
3042  Prerequisite: One of MATH 220 or MATH 221 or M...   
3043  Prerequisite: One of MATH 220 or MATH 221 or M...   
3044  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3373  May be repeated in the same or separate terms ...   
3374  May be repeated in the same or separate terms ...   
3375  May be repeated in the same or separate terms ...   
3376  May be repeated in the same or separate terms ...   
3377    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3040  Quantitative Reasoning II course.  ...              A   
3041  Quantitative Reasoning II course.  ...              A   
3042  Quantitative Reasoning II course.  ...              A   
3043  Quantitative Reasoning II course.  ...              P   
3044  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3373                                NaN  ...              A   
3374                                NaN  ...              A   
3375                                NaN  ...              A   
3376                                NaN  ...              A   
3377                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3040  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3041             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3042             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3043             Closed  Laboratory-Discussion       LBD   06:00 PM  07:50 PM   
3044               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
...                 ...                    ...       ...        ...       ...   
3373             Closed                 Online       E16   ARRANGED       NaN   
3374  Open (Restri

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [8]:
df_IS = df[df["Subject"]=="IS"]
df_IS

Year    Term YearTerm Subject  Number  \
6045  2023  Spring  2023-sp      IS     100   
6046  2023  Spring  2023-sp      IS     101   
6047  2023  Spring  2023-sp      IS     101   
6048  2023  Spring  2023-sp      IS     101   
6049  2023  Spring  2023-sp      IS     101   
...    ...     ...      ...     ...     ...   
6294  2023  Spring  2023-sp      IS     597   
6295  2023  Spring  2023-sp      IS     597   
6296  2023  Spring  2023-sp      IS     597   
6297  2023  Spring  2023-sp      IS     597   
6298  2023  Spring  2023-sp      IS     599   

                                                   Name  \
6045   Exploring the iSchool with a Human-Centered Lens   
6046               Introduction to Information Sciences   
6047               Introduction to Information Sciences   
6048               Introduction to Information Sciences   
6049               Introduction to Information Sciences   
...                                                 ...   
6294  Advanced Topics in Data Analytics &amp; Data S...   
6295  Advanced Topics in Data Analytics &amp; Data S...   
6296  Advanced Topics in Data Analytics &amp; Data S...   
6297  Advanced Topics in Data Analytics &amp; Data S...   
6298                                    Thesis Research   

                                            Description    Credit Hours  \
6045  This course introduces students to the School ...        1 hours.   
6046  This course provides an introduction to the fi...        3 hours.   
6047  This course provides an introduction to the fi...        3 hours.   
6048  This course provides an introduction to the fi...        3 hours.   
6049  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6294  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6295  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6296  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6297  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6298  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6045  Prerequisite: Restricted to Majors Only; First...   
6046                                                NaN   
6047                                                NaN   
6048                                                NaN   
6049                                                NaN   
...                                                 ...   
6294  2 or 4 graduate hours. No professional credit....   
6295  2 or 4 graduate hours. No professional credit....   
6296  2 or 4 graduate hours. No professional credit....   
6297  2 or 4 graduate hours. No professional credit....   
6298  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6045                                 NaN  ...              A   
6046  Social & Beh Sci - Soc Sci course.  ...              A   
6047  Social & Beh Sci - Soc Sci course.  ...              A   
6048  Social & Beh Sci - Soc Sci course.  ...              A   
6049  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6294                                 NaN  ...              A   
6295                                 NaN  ...              A   
6296                                 NaN  ...              A   
6297                                 NaN  ...              A   
6298                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6045  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6046             Closed  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6047               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6048             Closed  Disc

### 🔬 Test Case Checkpoint 🔬


In [9]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?


STAT offers the fewest course sections and CS offers the most course sections.


### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


They all offer a 599 numbered course. Across all three departments, this course is intended to be taken as part of a research thesis for graduate students.

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!



### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [10]:
mydf = df[df["Subject"]=="BIOE"]
mydf

Year    Term YearTerm Subject  Number  \
1586  2023  Spring  2023-sp    BIOE     120   
1587  2023  Spring  2023-sp    BIOE     200   
1588  2023  Spring  2023-sp    BIOE     202   
1589  2023  Spring  2023-sp    BIOE     202   
1590  2023  Spring  2023-sp    BIOE     202   
1591  2023  Spring  2023-sp    BIOE     202   
1592  2023  Spring  2023-sp    BIOE     202   
1593  2023  Spring  2023-sp    BIOE     205   
1594  2023  Spring  2023-sp    BIOE     210   
1595  2023  Spring  2023-sp    BIOE     297   
1596  2023  Spring  2023-sp    BIOE     303   
1597  2023  Spring  2023-sp    BIOE     303   
1598  2023  Spring  2023-sp    BIOE     303   
1599  2023  Spring  2023-sp    BIOE     306   
1600  2023  Spring  2023-sp    BIOE     306   
1601  2023  Spring  2023-sp    BIOE     310   
1602  2023  Spring  2023-sp    BIOE     360   
1603  2023  Spring  2023-sp    BIOE     397   
1604  2023  Spring  2023-sp    BIOE     414   
1605  2023  Spring  2023-sp    BIOE     415   
1606  2023  Spring  2023-sp    BIOE     415   
1607  2023  Spring  2023-sp    BIOE     415   
1608  2023  Spring  2023-sp    BIOE     415   
1609  2023  Spring  2023-sp    BIOE     415   
1610  2023  Spring  2023-sp    BIOE     416   
1611  2023  Spring  2023-sp    BIOE     416   
1612  2023  Spring  2023-sp    BIOE     420   
1613  2023  Spring  2023-sp    BIOE     435   
1614  2023  Spring  2023-sp    BIOE     436   
1615  2023  Spring  2023-sp    BIOE     479   
1616  2023  Spring  2023-sp    BIOE     481   
1617  2023  Spring  2023-sp    BIOE     481   
1618  2023  Spring  2023-sp    BIOE     484   
1619  2023  Spring  2023-sp    BIOE     484   
1620  2023  Spring  2023-sp    BIOE     486   
1621  2023  Spring  2023-sp    BIOE     486   
1622  2023  Spring  2023-sp    BIOE     487   
1623  2023  Spring  2023-sp    BIOE     497   
1624  2023  Spring  2023-sp    BIOE     498   
1625  2023  Spring  2023-sp    BIOE     499   
1626  2023  Spring  2023-sp    BIOE     500   
1627  2023  Spring  2023-sp    BIOE     570   
1628  2023  Spring  2023-sp    BIOE     570   
1629  2023  Spring  2023-sp    BIOE     572   
1630  2023  Spring  2023-sp    BIOE     572   
1631  2023  Spring  2023-sp    BIOE     574   
1632  2023  Spring  2023-sp    BIOE     574   
1633  2023  Spring  2023-sp    BIOE     575   
1634  2023  Spring  2023-sp    BIOE     575   
1635  2023  Spring  2023-sp    BIOE     580   
1636  2023  Spring  2023-sp    BIOE     588   
1637  2023  Spring  2023-sp    BIOE     597   
1638  2023  Spring  2023-sp    BIOE     598   
1639  2023  Spring  2023-sp    BIOE     598   
1640  2023  Spring  2023-sp    BIOE     598   
1641  2023  Spring  2023-sp    BIOE     598   
1642  2023  Spring  2023-sp    BIOE     598   
1643  2023  Spring  2023-sp    BIOE     598   
1644  2023  Spring  2023-sp    BIOE     599   

                                                   Name  \
1586                     Introduction to Bioengineering   
1587                    Bioengineering Career Immersion   
1588                  Cell &amp; Tissue Engineering Lab   
1589                  Cell &amp; Tissue Engineering Lab   
1590                  Cell &amp; Tissue Engineering Lab   
1591                  Cell &amp; Tissue Engineering Lab   
1592                  Cell &amp; Tissue Engineering Lab   
1593                  Signals &amp; Systems in Bioengrg   
1594         Linear Algebra for Biomedical Data Science   
1595                                   Individual Study   
1596                        Quantitative Physiology Lab   
1597                        Quantitative Physiology Lab   
1598                        Quantitative Physiology Lab   
1599                                 Biofabrication Lab   
1600                                 Biofabrication Lab   
1601            Computational Tools for Biological Data   
1602                   Transport &amp; Flow in Bioengrg   
1603                                   Individual Study   
1604                         Biomedical Instrumentation   
1605 

### Puzzle 4.2:

Select **five random courses** from your major, using `mydf` and pandas functions:

In [11]:
mydf.sample(n=5)

Year    Term YearTerm Subject  Number  \
1641  2023  Spring  2023-sp    BIOE     598   
1595  2023  Spring  2023-sp    BIOE     297   
1598  2023  Spring  2023-sp    BIOE     303   
1605  2023  Spring  2023-sp    BIOE     415   
1589  2023  Spring  2023-sp    BIOE     202   

                                   Name  \
1641                     Special Topics   
1595                   Individual Study   
1598        Quantitative Physiology Lab   
1605     Biomedical Instrumentation Lab   
1589  Cell &amp; Tissue Engineering Lab   

                                            Description   Credit Hours  \
1641  Subject offerings of new and developing areas ...  1 TO 4 hours.   
1595  Special project or reading activity. May be re...  1 TO 4 hours.   
1598  Experiments involving the modeling and measure...       2 hours.   
1605  Laboratory to accompany BIOE 414. Use of senso...       2 hours.   
1589  Principles of cell biology inherent in tissue ...       2 hours.   

                                           Section Info Degree Attributes  \
1641  1 to 4 graduate hours. No professional credit....               NaN   
1595  May be repeated in the same or separate terms ...               NaN   
1598  Prerequisite: Concurrent enrollment in BIOE 30...               NaN   
1605  Same as ECE 415. 2 undergraduate hours. 2 grad...               NaN   
1589  Prerequisite: MCB 150, and credit or concurren...               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
1641  ...              A  Open (Restricted)              Online       ONL   
1595  ...              A  Open (Restricted)   Independent Study       IND   
1598  ...              A  Open (Restricted)  Lecture-Discussion       LCD   
1605  ...              A             Closed          Laboratory       LAB   
1589  ...              A             Closed          Laboratory       LAB   

     Start Time  End Time Days of Week  Room            Building  \
1641   ARRANGED       NaN          NaN   NaN                 NaN   
1595   ARRANGED       NaN          NaN   NaN                 NaN   
1598   01:00 PM  01:50 PM            M  1306  Everitt Laboratory   
1605   01:00 PM  03:50 PM            M   201  Everitt Laboratory   
1589   12:00 PM  03:50 PM            R  3109  Everitt Laboratory   

                         Instructors  
1641                          Nie, S  
1595                             NaN  
1598                    Cvetkovic, C  
1605                         Reck, R  
1589  Gaj, T;Perez Pinera, P;Sirk, S  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered courses** from your major:

In [17]:
mydf.nsmallest(7,"Number")

Year    Term YearTerm Subject  Number  \
1586  2023  Spring  2023-sp    BIOE     120   
1587  2023  Spring  2023-sp    BIOE     200   
1588  2023  Spring  2023-sp    BIOE     202   
1589  2023  Spring  2023-sp    BIOE     202   
1590  2023  Spring  2023-sp    BIOE     202   
1591  2023  Spring  2023-sp    BIOE     202   
1592  2023  Spring  2023-sp    BIOE     202   

                                   Name  \
1586     Introduction to Bioengineering   
1587    Bioengineering Career Immersion   
1588  Cell &amp; Tissue Engineering Lab   
1589  Cell &amp; Tissue Engineering Lab   
1590  Cell &amp; Tissue Engineering Lab   
1591  Cell &amp; Tissue Engineering Lab   
1592  Cell &amp; Tissue Engineering Lab   

                                            Description Credit Hours  \
1586  Lectures and discussions of recent trends in b...     1 hours.   
1587  This course provides exposure to Bioengineerin...     1 hours.   
1588  Principles of cell biology inherent in tissue ...     2 hours.   
1589  Principles of cell biology inherent in tissue ...     2 hours.   
1590  Principles of cell biology inherent in tissue ...     2 hours.   
1591  Principles of cell biology inherent in tissue ...     2 hours.   
1592  Principles of cell biology inherent in tissue ...     2 hours.   

                                           Section Info Degree Attributes  \
1586                                                NaN               NaN   
1587               Prerequisite: BIOE 120. Majors only.               NaN   
1588  Prerequisite: MCB 150, and credit or concurren...               NaN   
1589  Prerequisite: MCB 150, and credit or concurren...               NaN   
1590  Prerequisite: MCB 150, and credit or concurren...               NaN   
1591  Prerequisite: MCB 150, and credit or concurren...               NaN   
1592  Prerequisite: MCB 150, and credit or concurren...               NaN   

      ... Section Status  Enrollment Status        Type Type Code Start Time  \
1586  ...              A  Open (Restricted)     Lecture       LEC   11:00 AM   
1587  ...              A  Open (Restricted)     Lecture       LEC   12:00 PM   
1588  ...              A             Closed  Laboratory       LAB   12:00 PM   
1589  ...              A             Closed  Laboratory       LAB   12:00 PM   
1590  ...              A               Open  Laboratory       LAB   04:00 PM   
1591  ...              A             Closed  Laboratory       LAB   11:00 AM   
1592  ...              A  Open (Restricted)     Lecture       LEC   12:00 PM   

      End Time Days of Week  Room                       Building  \
1586  11:50 AM           MW    35  Campus Instructional Facility   
1587  12:50 PM            W  2101             Everitt Laboratory   
1588  03:50 PM            T  3109             Everitt Laboratory   
1589  03:50 PM            R  3109             Everitt Laboratory   
1590  07:50 PM            T  3109             Everitt Laboratory   
1591  02:50 PM            M  3109             Everitt Laboratory   
1592  12:50 PM            F  2310             Everitt Laboratory   

                         Instructors  
1586                 Best-Popescu, C  
1587                    Vozenilek, J  
1588  Gaj, T;Perez Pinera, P;Sirk, S  
1589  Gaj, T;Perez Pinera, P;Sirk, S  
1590  Gaj, T;Perez Pinera, P;Sirk, S  
1591  Gaj, T;Perez Pinera, P;Sirk, S  
1592  Gaj, T;Perez Pinera, P;Sirk, S  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered courses** from your major:

In [18]:
mydf.nlargest(9,"Number")

Year    Term YearTerm Subject  Number  \
1644  2023  Spring  2023-sp    BIOE     599   
1638  2023  Spring  2023-sp    BIOE     598   
1639  2023  Spring  2023-sp    BIOE     598   
1640  2023  Spring  2023-sp    BIOE     598   
1641  2023  Spring  2023-sp    BIOE     598   
1642  2023  Spring  2023-sp    BIOE     598   
1643  2023  Spring  2023-sp    BIOE     598   
1637  2023  Spring  2023-sp    BIOE     597   
1636  2023  Spring  2023-sp    BIOE     588   

                                                   Name  \
1644                                    Thesis Research   
1638                                     Special Topics   
1639                                     Special Topics   
1640                                     Special Topics   
1641                                     Special Topics   
1642                                     Special Topics   
1643                                     Special Topics   
1637                                   Individual Study   
1636  Biomedical Image Computing Capstone Project Li...   

                                            Description    Credit Hours  \
1644  Bioengineering graduate thesis research. Appro...  0 TO 16 hours.   
1638  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1639  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1640  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1641  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1642  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1643  Subject offerings of new and developing areas ...   1 TO 4 hours.   
1637  Special project or reading activity. May be re...   1 TO 8 hours.   
1636  This journal-club style course will expose stu...        1 hours.   

                                           Section Info Degree Attributes  \
1644    Approved for S/U grading only. May be repeated.               NaN   
1638  1 to 4 graduate hours. No professional credit....               NaN   
1639  1 to 4 graduate hours. No professional credit....               NaN   
1640  1 to 4 graduate hours. No professional credit....               NaN   
1641  1 to 4 graduate hours. No professional credit....               NaN   
1642  1 to 4 graduate hours. No professional credit....               NaN   
1643  1 to 4 graduate hours. No professional credit....               NaN   
1637  May be repeated. Prerequisite: Approved writte...               NaN   
1636  1 graduate hour. No professional credit. Appro...               NaN   

      ... Section Status           Enrollment Status                Type  \
1644  ...              A           Open (Restricted)   Independent Study   
1638  ...              A           Open (Restricted)             Lecture   
1639  ...              A                      Closed             Lecture   
1640  ...              A                      Closed  Lecture-Discussion   
1641  ...              A           Open (Restricted)              Online   
1642  ...              A  CrossListOpen (Restricted)             Lecture   
1643  ...              A           Open (Restricted)              Online   
1637  ...              A           Open (Restricted)   Independent Study   
1636  ...              A           Open (Restricted)  Lecture-Discussion   

     Type Code Start Time  End Time Days of Week  Room  \
1644       IND   ARRANGED       NaN          NaN   NaN   
1638        L1   ARRANGED       NaN          NaN   NaN   
1639       LEC   03:30 PM  04:50 PM           TR  1302   
1640       LCD   05:00 PM  06:20 PM           TR  1306   
1641       ONL   ARRANGED       NaN          NaN   NaN   
1642       LEC   09:30 AM  10:50 AM           TR  2101   
1643       ONL   ARRANGED       NaN          NaN   NaN   
1637       IND   ARRANGED       NaN          NaN   NaN   
1636       LCD   10:00 AM  10:50 AM            F  2036   

                           Building    Instructors  
1644                       

### Puzzle 4.5:

Select **a random sample of 10% of the courses** from your major:

In [20]:
mydf.sample(frac=0.1)

Year    Term YearTerm Subject  Number  \
1636  2023  Spring  2023-sp    BIOE     588   
1598  2023  Spring  2023-sp    BIOE     303   
1603  2023  Spring  2023-sp    BIOE     397   
1642  2023  Spring  2023-sp    BIOE     598   
1612  2023  Spring  2023-sp    BIOE     420   
1625  2023  Spring  2023-sp    BIOE     499   

                                                   Name  \
1636  Biomedical Image Computing Capstone Project Li...   
1598                        Quantitative Physiology Lab   
1603                                   Individual Study   
1642                                     Special Topics   
1612                          Intro Bio Control Systems   
1625                                      Senior Thesis   

                                            Description   Credit Hours  \
1636  This journal-club style course will expose stu...       1 hours.   
1598  Experiments involving the modeling and measure...       2 hours.   
1603  Special project or reading activity. May be re...  1 TO 4 hours.   
1642  Subject offerings of new and developing areas ...  1 TO 4 hours.   
1612  Systems engineering approach to modeling physi...       3 hours.   
1625  Limited in general to seniors in the curriculu...  1 TO 5 hours.   

                                           Section Info Degree Attributes  \
1636  1 graduate hour. No professional credit. Appro...               NaN   
1598  Prerequisite: Concurrent enrollment in BIOE 30...               NaN   
1603  May be repeated up to 8 hours in a term to a m...               NaN   
1642  1 to 4 graduate hours. No professional credit....               NaN   
1612  3 undergraduate hours. No graduate credit. Cre...               NaN   
1625  1 to 5 undergraduate hours. No graduate credit...               NaN   

      ... Section Status           Enrollment Status                Type  \
1636  ...              A           Open (Restricted)  Lecture-Discussion   
1598  ...              A           Open (Restricted)  Lecture-Discussion   
1603  ...              A           Open (Restricted)   Independent Study   
1642  ...              A  CrossListOpen (Restricted)             Lecture   
1612  ...              A                      Closed             Lecture   
1625  ...              A                      Closed   Independent Study   

     Type Code Start Time  End Time Days of Week  Room  \
1636       LCD   10:00 AM  10:50 AM            F  2036   
1598       LCD   01:00 PM  01:50 PM            M  1306   
1603       IND   ARRANGED       NaN          NaN   NaN   
1642       LEC   09:30 AM  10:50 AM           TR  2101   
1612       LEC   11:00 AM  11:50 AM          MWF   119   
1625       IND   ARRANGED       NaN          NaN   NaN   

                           Building   Instructors  
1636  Campus Instructional Facility  Anastasio, M  
1598             Everitt Laboratory  Cvetkovic, C  
1603                            NaN           NaN  
1642             Everitt Laboratory       Amos, J  
1612    Materials Science & Eng Bld     Monroy, G  
1625                            NaN           NaN  

[6 rows x 27 columns]

### ✔️ Checkpoint ✔️

There is no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is still a new relatively new major here at UIUC. The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [21]:
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the 2023 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [22]:
df[df["Subject"]== "IS"]

Year    Term YearTerm Subject  Number  \
6045  2023  Spring  2023-sp      IS     100   
6046  2023  Spring  2023-sp      IS     101   
6047  2023  Spring  2023-sp      IS     101   
6048  2023  Spring  2023-sp      IS     101   
6049  2023  Spring  2023-sp      IS     101   
...    ...     ...      ...     ...     ...   
6294  2023  Spring  2023-sp      IS     597   
6295  2023  Spring  2023-sp      IS     597   
6296  2023  Spring  2023-sp      IS     597   
6297  2023  Spring  2023-sp      IS     597   
6298  2023  Spring  2023-sp      IS     599   

                                                   Name  \
6045   Exploring the iSchool with a Human-Centered Lens   
6046               Introduction to Information Sciences   
6047               Introduction to Information Sciences   
6048               Introduction to Information Sciences   
6049               Introduction to Information Sciences   
...                                                 ...   
6294  Advanced Topics in Data Analytics &amp; Data S...   
6295  Advanced Topics in Data Analytics &amp; Data S...   
6296  Advanced Topics in Data Analytics &amp; Data S...   
6297  Advanced Topics in Data Analytics &amp; Data S...   
6298                                    Thesis Research   

                                            Description    Credit Hours  \
6045  This course introduces students to the School ...        1 hours.   
6046  This course provides an introduction to the fi...        3 hours.   
6047  This course provides an introduction to the fi...        3 hours.   
6048  This course provides an introduction to the fi...        3 hours.   
6049  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6294  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6295  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6296  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6297  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6298  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6045  Prerequisite: Restricted to Majors Only; First...   
6046                                                NaN   
6047                                                NaN   
6048                                                NaN   
6049                                                NaN   
...                                                 ...   
6294  2 or 4 graduate hours. No professional credit....   
6295  2 or 4 graduate hours. No professional credit....   
6296  2 or 4 graduate hours. No professional credit....   
6297  2 or 4 graduate hours. No professional credit....   
6298  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6045                                 NaN  ...              A   
6046  Social & Beh Sci - Soc Sci course.  ...              A   
6047  Social & Beh Sci - Soc Sci course.  ...              A   
6048  Social & Beh Sci - Soc Sci course.  ...              A   
6049  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6294                                 NaN  ...              A   
6295                                 NaN  ...              A   
6296                                 NaN  ...              A   
6297                                 NaN  ...              A   
6298                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6045  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6046             Closed  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6047               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6048             Closed  Disc

In [23]:
df2019[df2019["Subject"]== "IS"]

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [ ]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

In [24]:
254-92

162

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

By analyzing the number of rows in both the 2019 and current datasets, I found that the program grew by 162 courses!

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)